File: 509-fibonacci-number/509-fibonacci-number.py
class Solution:
    def fib(self, n: int, memo = {}) -> int:
        if n in memo: return memo[n]
        if n < 2: return n
        memo[n] = self.fib(n-1, memo) + self.fib(n-2,memo)
        return memo[n]
        

File: 509-fibonacci-number/509-fibonacci-number.java
class Solution {
    public int fib(int n) {
        if (n <= 1){
            return n;
        }
        int prev1 = 0;
        int prev2 = 1;
        for (int i = 2; i <= n; i++){
            int temp = prev2;
            prev2 += prev1;
            prev1 = temp;
        }
        return prev2;
    }
}